tests/product/image_builder.py

```diff
@@ -0,0 +1,171 @@
+                                    StandaloneModeInstaller,
+                                    TopologyInstaller,
+                                    StandalonePrestoInstaller],
+        PA_ONLY_YS_CLUSTER: [PrestoadminInstaller,
+                             YarnSliderModeInstaller],
+        PA_SLIDER_CLUSTER: [PrestoadminInstaller,
+                            YarnSliderModeInstaller,
+                            SliderInstaller]
+    }
+
+    def __init__(self, caller):
+        self.caller = caller
+        self.caller.default_keywords = {}
+        self.caller.cluster = None
+
+    def _apply_post_install_hooks(self, installers):
+        for installer in installers:
+            self.caller.cluster.postinstall(installer)
+
+    def _update_replacement_keywords(self, installers):
+        for installer in installers:
+            installer_instance = installer(self.caller)
+            self.caller.default_keywords.update(installer_instance.get_keywords())
+
+    def _run_installers(self, installers):
+        cluster = self.caller.cluster
+        for installer in installers:
+            dependencies = installer.get_dependencies()
+
+            for dependency in dependencies:
+                dependency.assert_installed(self.caller)
+
+            installer_instance = installer(self.caller)
+            installer_instance.install()
+
+            self.caller.default_keywords.update(installer_instance.get_keywords())
+            cluster.postinstall(installer)
+
+    def _setup_cluster(self, bare_image_provider, cluster_type):
+        installers = self._cluster_types[cluster_type]
+
+        config_filename = ConfigurableCluster.check_for_cluster_config()
+
+        if config_filename:
+            self.caller.cluster = ConfigurableCluster.start_bare_cluster(
+                config_filename, self.caller,
+                StandalonePrestoInstaller.assert_installed)
+        else:
+            self.caller.cluster, bare_cluster = DockerCluster.start_cluster(
+                bare_image_provider, cluster_type)
+
+            # If we've found images and started a non-bare cluster, the
+            # containers have already had the installers applied to them.
+            # We do need to get the test environment in sync with the
+            # containers by calling the following two functions.
+            #
+            # Once that's done, the cluster and test environment is in the
+            # same state it would be as if we'd called _run_installers on
+            # a bare cluster, and we can return.
+            #
+            # We do this to save the cost of running the installers on the
+            # docker containers every time we run a test. In practice,
+            # that turns out to be a fairly expensive thing to do.
+            if not bare_cluster:
+                self._apply_post_install_hooks(installers)
+                self._update_replacement_keywords(installers)
+                return self.caller.cluster
```

Comment:
I believe these need to be called every time we create a new cluster from existing images.

If that's no longer true, they shouldn't be called here, because we should always have a bare cluster when we get there.